### ISSUE TITLE
[BUG] Actor Naming does not respect " - "

### ISSUE THREAD

**Issue by andar1an** [wasmCloud/wash-archive#341]
**Describe the bug**
Create an actor with a dash in the name, and try to build. 
```Could not find compiled wasm file to sign: target/wasm32-unknown-unknown/release/actor-name.wasm```

**To Reproduce**
1. Run command `wash new actor actor-name -t hello`
1. Run other command `wash build`
1. See error

**Expected behavior**
Wash respects dash in an actor name, and build without needing to change the name in wasmcloud.toml to use an underscore (@jordan-rash discovered this workaround).

**Screenshots**

**Environment (please complete the following information) **
 - OS: Manjaro 22.0.0
 - Shell: zsh
 - wash Version: 13

**Comment by stevelr** [comment]
Some additional background information and how I think we could fix it:

There are a couple limitations that restrict how we can handle hyphens:
- cargo won't compile a library name with a hyphen in it, so the project name needs to be changed to have an underscore. The generated wash file name will have an underscore in the name. 
- it is possible to have a actor name that contains a dash, even though the wasm file contains underscores, by using the `--name` parameter to `wash sign`. That would let the actor show up in host inventory and the washboard with a hyphen, but that's a bit limited: Call aliases also can't have underscores (currently), so if you were planning to call the actor with a friendly name you'd have to change the hyphens to underscores. 

Options:

(1) replace the project name with underscore, so at least wash new && wash build don't fail.  This can be done with a change to the project templates: Use  `name = {{to_snake_case project-name}}` (whoever implements this, it should be done for all project templates). We may want to print a notice on wash's output to tell the user that the name was changed. 

(2) change the project name and also rename the actor at signing time. In addition to the change (1) on the template, this would require a change to the wasmcloud.toml parser and adding the `--name` parameter to `wash sign`.

(3) reject project names with hyphens and make the user re-submit the wash new command with a non-hyphenated name. 

IMO (1) is preferable to (3) since (3) is an abrupt failure and (1) succeeds with a minor change, plus a notice to train the user how to do it right the next time.

**Comment by andar1an** [comment]
I also like (1), especially with notice and auto-renaming to underscore.

**Comment by brooksmtownsend** [comment]
Agreed with above, `1` sounds like a built-in way to handle the problem!

**Comment by ricochet** [comment]
Another case to be handled that seems related:

```bash
wash new actor blobs3
cd blobs3
wash build
```

Compiles but then this error:
> Could not find compiled wasm file, please ensure target/wasm32-unknown-unknown/release/blobs3.wasm exists

```bash
ls target/wasm32-unknown-unknown/release
blobs_3.wasm 
```

Something added an underscore because the name is postfixed with a number.

**Comment by andar1an** [comment]
> Another case to be handled that seems related:
> 
> ```bash
> wash new actor blobs3
> cd blobs3
> wash build
> ```
> 
> Compiles but then this error:
> > Could not find compiled wasm file, please ensure target/wasm32-unknown-unknown/release/blobs3.wasm exists
> 
> ```bash
> ls target/wasm32-unknown-unknown/release
> blobs_3.wasm 
> ```
> 
> Something added an underscore because the name is postfixed with a number.

I have also experienced this and have had to manually rename. I did not know if it was by design. Affects providers and interfaces as well I believe.

**Comment by brooksmtownsend** [comment]
Hm, if we have an idea for what all the rules are for dashes, numbers, and underscores, we can start building them into our project templates. The edge case with numbers wasn't something I was aware of 😕

**Comment by autodidaddict** [comment]
We should clarify that we're talking about the cargo project name, right? Because the "name" of an actor when embedded in a JWT is just a string with no restrictions

**Comment by andar1an** [comment]
The problem is evident for numbers in more than 1 place, but they all have to do with files that I have seen. I personally have not looked at JWT.

**Comment by vados-cosmonic** [comment]
Hey all I think I want to take a stab at this -- as I see it, the issue is that no 

But this makes me think of a few more things:

Right now we only use a single template name -- can we explicitly specify the *programming language* the template is for? As what actors can be called is not limited, the restriction we're running into is actually language specific, and we don't have a transparent way to represent that complexity.

Rather than trying to solve the problem at the template level I think it might be a good idea to change the name at the wash level, print the message *then* go about the rest of the process as normal?

I've got a [sketch of what I'd like to do here](https://github.com/wasmCloud/wash/pull/390) -- there's some room for improvement but I think it's a good starting point for discussion if there is any!

**Comment by ghost** [comment]
> can we explicitly specify the programming language the template is for?

IMO this gets very complicated very quickly especially as the number of supported languages grows and they all have their own naming conventions. I think we just need to define what the convention is for a wasmCloud actor (and provider?) name. The most obvious pick is `snake_case` given that most actors right now are in Rust, so it would cause the least churn.

Regardless of what the actual convention is, after we pick one, the user would just need to create a name that matches the wasmCloud naming convention. If a supported language requires us to write the name differently (ex. if we supported Java it would be `CamelCase`) we just make that transformation for the language-specific files of that particular template, and keep the actual actor name the same.

**Comment by jordan-rash** [comment]
I'm working on templates for Go providers and more actors right now and most everything is [MixedCaps](https://go.dev/doc/effective_go#mixed-caps)

**Comment by vados-cosmonic** [comment]
@mattwilkinsonn Hmm, yeah going with snake case feels like it would work for *most* other languages. I want to avoid over-specializing on Rust because `wash` should enable all WASM-ready languages, eventually... 

I personally think that knowing the target language of the template/what you're generating for is useful and would open doors in the future (as templates proliferate, what if we want to limit which are shown by language?) 

Can we do a bit of both? There's an `Unknown` case in there where most things can fall in, and we can require snake case as the default there, but make language specific changes when necessary. I'm a bit worried about painting ourselves into a corner and having to make a change to the naming scheme for some other language later. 

Let's say we did go for one standardized naming scheme -- One case I haven't covered yet is the the number issue (that `blobs3` example) -- it looks to be a rust-only thing, so should we add this rule for all languages?

The Golang example brought up by @jordan-rash is interesting -- gophers use MixedCaps names and likely the template might expect a MixedCaps name to come through? This means people who write templates need to ensure the case is MixedCaps (to they'd need to use a helper in their template to ensure conformity (don't know if it's an error to name it differently there).

I'm generally a fan of very explicit behavior, a kind of add-on choice we're making is whether to actually separate the wasmcloud actor name and the generated module name, should we make that explicit? Printing a message is good, but generating a module name to the best of our ability (but including a way to override it specifically) might be good.

**Comment by ghost** [comment]
@vados-cosmonic

My pitch is that the wasmCloud actor name is the "source of truth", and what the template names its' files and puts in the language config files is an implementation detail. So we would never change the wasmCloud naming scheme for different languages, Rust or not. When using a template, the actor/provider name can be transformed into a version friendly to the language for all of the language specific files and configs, but remain as the original version in `wasmcloud.toml`.

I'm also afraid of adding too many customization options to the templates for the sake of both the template authors and the end users. As long as we can transform all valid wasmCloud names to a version that is accepted by the language and successfully compiles, we don't need any customization options for module/file names. If the user then wants to modify the module/package/crate/wheel (etc etc) name or the generated file name(s) to something else, that's totally fine, they are then taking on the overhead of making sure it still compiles with their changes.

**Comment by vados-cosmonic** [comment]
(excuse the length! I did a bit of recapping just to make sure I'm 100% on the same page as everyone else)

> My pitch is that the wasmCloud actor name is the "source of truth", and what the template names its' files and puts in the language config files is an implementation detail. So we would never change the wasmCloud naming scheme for different languages, Rust or not. When using a template, the actor/provider name can be transformed into a version friendly to the language for all of the language specific files and configs, but remain as the original version in wasmcloud.toml.

I agree with this, but the problem is right now there is no wasmCloud actor naming scheme -- this issue has evolved to *to introduce one*, right? 

Maybe I am misunderstanding your point -- would you like to keep the wasmCloud side naming completely unfiltered (just a string, with no limitations AFAIK, with comments in this issue), and leave it up to template writers to make all necessary changes per-language? If this is the case, then there's no need to change `wash` at all, right?

What I thought we were discussing is that since `wash`'s functionality is affected (i.e. when we try to build & run the generated WASM) by language toolkit choices, we needed to be more strict about what we allowed because language toolkits (in this limited case Rust) can change naming underneath us, and in addition be broken by our choices. 

So far we identified the following issues with *not* enforcing a naming scheme:

- replacing hyphens with underscores
- replacing places where numbers are next to letters with `_<num>`

Taking a step back -- we can either solve this in `wash`, or leave it up to template writers.

These above issues are rust-specific (at present) so my proposal is a language-dependent solution. We have the option of defining a standard that is likely to work across languages, by severely restricting the subset of possible names, which is fine too.

At the high level, there is one crucial question and one follow on:

1. Do we make an attempt to solve this in `wash`, or leave it to template writers to take the unfiltered string?
2. (optional) If we've chosen to solve it in `wash`, do we make the changes language-specific or the common denominator that's likely to work across all languages (underscores over hyphens + some specific changes like the number thing)

IMO If `wash` takes the input, the program should build & run. `wash` should do the work to make sure that future `wash`steps work properly.

> I'm also afraid of adding too many customization options to the templates for the sake of both the template authors and the end users. As long as we can transform all valid wasmCloud names to a version that is accepted by the language and successfully compiles, we don't need any customization options for module/file names.

Well on this issue I think I'm a little bit on the other side -- well used customization enables flexibility for both users and template authors. There can be vastly different ways to template out code (ex. should some generated rust code use `libx` or `liby`?), and as an intermediary `wash` should be negotiating this/doing interchange, so users get what they want and template authors get the information they need to function properly.

That aside though, I agree that this issue does *not* need to include customization options for module names -- my point was that making the difference explicit might be beneficial but we can decide that in the future when someone walks up with a use case/need.

>  If the user then wants to modify the module/package/crate/wheel (etc etc) name or the generated file name(s) to something else, that's totally fine, they are then taking on the overhead of making sure it still compiles with their changes.

I think `wash` (and transiently template authors) ought to take some more responsibility here -- `wash` exists to save users time (as opposed to just pulling down a zip file and going in and making edits), and IMO offering clarity/removing the need to fuss around with files after templating is within the purview. 

With regards to your point, I do agree that we can't include *every* kind of non-functional change in `wash` as configuration, but the problem is that this particular issue currently breaks `wash`.

**Comment by ghost** [comment]
@vados-cosmonic 

Quick preface - I'm not beholden to this idea of having a enforced naming scheme. I just think that it would solve these issues and make it easy to define what we do in situations where style conventions don't match up in the future. Now to your comments:

> I agree with this, but the problem is right now there is no wasmCloud actor naming scheme -- this issue has evolved to to introduce one, right?

Yeah, at least I am of the opinion that we should.

> Maybe I am misunderstanding your point -- would you like to keep the wasmCloud side naming completely unfiltered (just a string, with no limitations AFAIK, with comments in this issue), and leave it up to template writers to make all necessary changes per-language? If this is the case, then there's no need to change wash at all, right?

No - i'm thinking that the wasmCloud names have a specific style that is enforced by `wash`. This can be `snake_case` or `CamelCase` or any other convention, I just think we should pick one and stick to it. It resolves issues like the original one from this issue and also makes it easier for people to find/type actor & provider names if they know what pattern all of them will follow.

> What I thought we were discussing is that since wash's functionality is affected (i.e. when we try to build & run the generated WASM) by language toolkit choices, we needed to be more strict about what we allowed because language toolkits (in this limited case Rust) can change naming underneath us, and in addition be broken by our choices.

Agreed on this. My resolution to this is that: If we have a specific style that we're conforming to, it's easy to then transform all names of that style to another one. 

>Do we make an attempt to solve this in wash, or leave it to template writers to take the unfiltered string?
>(optional) If we've chosen to solve it in wash, do we make the changes language-specific or the common denominator that's likely to work across all languages (underscores over hyphens + some specific changes like the number thing)

The template specifies what style transformation they want, IMO. We already do this in some places: https://github.com/wasmCloud/project-templates/blob/main/actor/echo/Cargo.toml

>IMO If wash takes the input, the program should build & run. wash should do the work to make sure that future washsteps work properly.

Agreed. If we have a specific naming convention then we can logically ensure that all names from that convention can be transformed into valid names for the language. Generally this would just be by going from `snake_case` to `UpperCamelCase` or `lowerCamelCase` in most cases. I can't think of another transformation we'd need to do at the moment.

> Well on this issue I think I'm a little bit on the other side -- well used customization enables flexibility for both users and template authors. There can be vastly different ways to template out code (ex. should some generated rust code use libx or liby?), and as an intermediary wash should be negotiating this/doing interchange, so users get what they want and template authors get the information they need to function properly.

I don't think templates should be going this far. They are primarily a way to get people started on wasmCloud development so they can then learn the ropes, so should be kept as simple as possible with as small a setup time as possible. Advanced changes (like, IMO, your customizing generated code libraries example), should just be something a user can do after the template has been created, along with documentation on how to do it.

This is a bit of a philosophical difference though - I personally dislike template "wizards". I'd rather the command do as little as possible to accomplish the task and then link me to documentation that tells me ways I can change and modify things later.

> I think wash (and transiently template authors) ought to take some more responsibility here -- wash exists to save users time (as opposed to just pulling down a zip file and going in and making edits), and IMO offering clarity/removing the need to fuss around with files after templating is within the purview.

I'm envisaging most users just copy/pasting a command that pulls down a template and running it. Adding a feature that renames files is just kinda overkill IMO, for our implementation side and for the user to use. Specifically for the user - Why go figure out what CLI flag you need to pass, or deal with an interactive CLI prompt, when you could just create the template and then rename the file(s) in the way that you are used to?

**Comment by vados-cosmonic** [comment]
Thanks for the detailed response, I think we're mostly in agreement on actual implementation here, but I want to make it really explicit the choice being made, so a bit more rehash:

> No - i'm thinking that the wasmCloud names have a specific style that is enforced by wash. This can be snake_case or CamelCase or any other convention, I just think we should pick one and stick to it. It resolves issues like the original one from this issue and also makes it easier for people to find/type actor & provider names if they know what pattern all of them will follow.
> 
> Agreed on this. My resolution to this is that: If we have a specific style that we're conforming to, it's easy to then transform all names of that style to another one.
>
> Agreed. If we have a specific naming convention then we can logically ensure that all names from that convention can be transformed into valid names for the language. Generally this would just be by going from snake_case to UpperCamelCase or lowerCamelCase in most cases. I can't think of another transformation we'd need to do at the moment.

Leaning into these points, the proposal sounds like it would shape up to the following *in practice*:

- all lower case
- underscores as separation (no hyphens)
- no numbers next to letters (`blob3` case)

Does that sound right to you? The last point is rust-specific, but it currently breaks `wash` -- we either add it as a rule, or require every template to include a some sort of `{{space_numbers_with_underscore ...}}` handlebars helper.

This ruleset represents what we know to be issues and likely tolerable right now, but even if this were the solution, **there's a problem, here's the scenario**:

- `wash new` takes `MixedCaps3` as input, alerts and changes to `mixed_caps_3`
- golang template takes `mixed_caps_3` and turns it into `MixedCaps3` (or `MixedCaps3.wasm`, `MixedCapsThree.wasm` etc)
- `wash` crashes because [it expected `mixed_caps_3.wasm` (the name in common config)](https://github.com/wasmCloud/wash/blob/main/crates/wash-lib/src/build.rs#L145) to be the name of the generated executable

The core problem is that `wash` needs to know (optionally, control) the resulting generated wasm module name. If the template changes that out from under us, wash needs to find it out somehow, but this requires *not* allowing template writers to do anything they want with it.

In fact, perusing some of the `wash-lib` code encourages the language-dependent leaning I had -- we're already [dispatching based on language specifics](https://github.com/wasmCloud/wash/blob/main/crates/wash-lib/src/build.rs#L78).

<details>
<summary><h4>Other things worth saying, but less important than the above</h4></summary>

> The template specifies what style transformation they want, IMO. We already do this in some places: https://github.com/wasmCloud/project-templates/blob/main/actor/echo/Cargo.toml

I did see that -- but my thing is that the choices we make either add or remove work for tens/hundreds of template maintainers, so I'd err on the side of putting in the work to have something that "just works". 

Also as a reference, `cargo` acts an example -- they solve this by accepting all applicable crate-level names and leaving an override point for library names.

> I don't think templates should be going this far. They are primarily a way to get people started on wasmCloud development so they can then learn the ropes, so should be kept as simple as possible with as small a setup time as possible. Advanced changes (like, IMO, your customizing generated code libraries example), should just be something a user can do after the template has been created, along with documentation on how to do it.
>
> This is a bit of a philosophical difference though - I personally dislike template "wizards". I'd rather the command do as little as possible to accomplish the task and then link me to documentation that tells me ways I can change and modify things later.
> 
> I'm envisaging most users just copy/pasting a command that pulls down a template and running it. Adding a feature that renames files is just kinda overkill IMO, for our implementation side and for the user to use. Specifically for the user - Why go figure out what CLI flag you need to pass, or deal with an interactive CLI prompt, when you could just create the template and then rename the file(s) in the way that you are used to?

I see your point, and I think we look at this differently. I see this as the end-user copying or pasting one command versus two.

To expand: from whatever documentation/blogpost/resource the end-user reads, they either copy and paste *one* command with toggles to get the functionality they want, or they copy paste one command then do <1-n extra commands> to get up to a ready-to-go state with the template they want.

With proper customization outlets, it is possible for template providers who are obsessed with smooth user experience to write templates that do the right thing in one properly toggled invocation. Without these affordances, it is impossible, and the user *must* do extra work. While I agree that minimalism in tooling is preferred, requiring more work by default seems like leaving potential ux gains on the table. 

That said, I agree it's a philosophical argument that we can (and should) punt to later so I'm happy to punt it here.
</details>

**Comment by ghost** [comment]
I've titled each section here to help us keep track of the different discussions.

**Example scenario**
>This ruleset represents what we know to be issues and likely tolerable right now, but even if this were the solution, there's a problem, here's the scenario:
> - wash new takes MixedCaps3 as input, alerts and changes to mixed_caps_3
> - golang template takes mixed_caps_3 and turns it into MixedCaps3 (or MixedCaps3.wasm, MixedCapsThree.wasm etc)
> - wash crashes because [it expected mixed_caps_3.wasm (the name in common config)](https://github.com/wasmCloud/wash/blob/main/crates/wash-lib/src/build.rs#L145) to be the name of the generated executable

This isn't how it works - when compiling a TinyGo actor, we explicitly tell the compiler where we want it to put the output by passing a `-o` flag, which we fill in with the `wasmcloud.toml` name. [Relevant `wash` code](https://github.com/wasmCloud/wash/blob/main/crates/wash-lib/src/build.rs#L172-L190) and [TinyGo docs](https://tinygo.org/docs/reference/usage/important-options/).

So in your example, there would never be a `MixedCaps3.wasm` file, it will always be `mixed_caps_3.wasm`.

Let me know if this makes sense or I'm missing a wrinkle here!

**Style proposal**
>Leaning into these points, the proposal sounds like it would shape up to the following in practice:
> - all lower case
> - underscores as separation (no hyphens)
> -no numbers next to letters (blob3 case)

Yeah, that sounds reasonable to me. However we should probably make this as a seperate RFC to narrow down what the exact style is. The 3rd point is probably one people might have feedback on, I'd personally lean towards not including it and just having a handlebar transform for all Rust-specific names, but I'm not overly worried about which path we go down for that.

**Knowing the wasm module names**
> The core problem is that wash needs to know (optionally, control) the resulting generated wasm module name. If the template changes that out from under us, wash needs to find it out somehow, but this requires not allowing template writers to do anything they want with it.

It already does, at least for Go. [This code for Rust](https://github.com/wasmCloud/wash/blob/main/crates/wash-lib/src/build.rs#L137-L146) is actually buggy, if the `wasmcloud.toml` name and crate name didn't match up it would fail to find the compiled actor, which is actually the cause of the original issue here 😂. Need to change this to fetch the crate's name directly from the `Cargo.toml` file.


**Dispatch based on language**
> In fact, perusing some of the wash-lib code encourages the language-dependent leaning I had -- we're already [dispatching based on language specifics](https://github.com/wasmCloud/wash/blob/main/crates/wash-lib/src/build.rs#L78).

We have to do that on some level in the control flow because Rust and Go don't have the same compiler. That's why we split up them up there, and then we just hand the path of the compiled WASM actor to the rest of the function.

**Transforming names at `wash` level or template level**
>> The template specifies what style transformation they want, IMO. We already do this in some places: https://github.com/wasmCloud/project-templates/blob/main/actor/echo/Cargo.toml
>I did see that -- but my thing is that the choices we make either add or remove work for tens/hundreds of template maintainers, so I'd err on the side of putting in the work to have something that "just works".
>Also as a reference, cargo acts an example -- they solve this by accepting all applicable crate-level names and leaving an override point for library names.

I don't think it is difficult for a template author to write `{{to_camel_case ...}}` or similar when they want to transform the name. I think that also allows for greater flexibility when creating templates, and makes it clear to the users of the template what is happening. Also I don't think `cargo` is a good comp because it doesn't deal with multiple language ecosystems. Something like [Bazel](https://bazel.build/) or another cross-language build toolchain may be useful to compare here.

**Templates**
> I see your point, and I think we look at this differently. I see this as the end-user copying or pasting one command versus two.
> To expand: from whatever documentation/blogpost/resource the end-user reads, they either copy and paste one command with toggles to get the functionality they want, or they copy paste one command then do <1-n extra commands> to get up to a ready-to-go state with the template they want.
> With proper customization outlets, it is possible for template providers who are obsessed with smooth user experience to write templates that do the right thing in one properly toggled invocation. Without these affordances, it is impossible, and the user must do extra work. While I agree that minimalism in tooling is preferred, requiring more work by default seems like leaving potential ux gains on the table.
> That said, I agree it's a philosophical argument that we can (and should) punt to later so I'm happy to punt it here.

I think boils down to that we disagree on what a template should provide. I'm seeing it as setting up the minimal things needed to build a wasmCloud actor. In my eyes, 100% of the time the user should be able to create a new actor from the template and then immediately run `wash build` and see it working. 

If we add UX features / code transformations / dependency swaps etc to this process the chances of something breaking multiplies with each one. This is really bad because when something does break, the user never even got to the "success" state, and then they just come away with an impression that `wash` is a bad tool that generates broken code.  If this is a first-time wasmCloud user, we could lose them for good off of something like this.

Instead, we give them as small a template as possible that we are certain always compiles right off the bat. They create their actor from the template, run `wash build`, and see a successfully compiled and signed actor. Inside the templates we can add comments about stuff they could change and also link them to documentation in the CLI output to encourage them to add & change whatever they'd like to.

If they then want to go rename some files, change out a dependency, or follow the documentation's guide to set up `foobar` -  they know that if the actor fails to compile afterwards, it was because of the specific change they made. This also puts them into a typical software development pattern of having a working build, changing something, seeing it break, and then figuring out what broke and how they can fix it.

Sorry, that was a lot of text! Let me know what you think on this.

**Comment by vados-cosmonic** [comment]
Thanks for the detailed response -- I've combined some sections since I think they're related, and to keep the issue-related discussion a bit tighter:

**Example scenario** / **Dispatch based on language**

> This isn't how it works - when compiling a TinyGo actor, we explicitly tell the compiler where we want it to put the output by passing a -o flag, which we fill in with the wasmcloud.toml name. [Relevant wash code](https://github.com/wasmCloud/wash/blob/main/crates/wash-lib/src/build.rs#L172-L190) and [TinyGo docs](https://tinygo.org/docs/reference/usage/important-options/).
>
> So in your example, there would never be a MixedCaps3.wasm file, it will always be mixed_caps_3.wasm.
>
> Let me know if this makes sense or I'm missing a wrinkle here!
>
> It already does, at least for Go. [This code for Rust](https://github.com/wasmCloud/wash/blob/main/crates/wash-lib/src/build.rs#L137-L146) is actually buggy, if the wasmcloud.toml name and crate name didn't match up it would fail to find the compiled actor, which is actually the cause of the original issue here :joy:. Need to change this to fetch the crate's name directly from the Cargo.toml file.

Thanks for informing me on the difference in how TinyGo works versus Rust -- this actually changes my understanding of `wash` drastically!

What I realize now is that the (slightly unequally held) implicit assumption is *`wash` don't trust the actor name to persist, if it can't control the build process* -- so assuming that our actor name has anything to do with the generated wasm module name in the first place was wrong.

Reading your comment I was originally thinking I should do two things:
- Write an RFC for pinning down actor naming rules
- Implement the ruleset as discussed

But it looks like what I should *actually* to solve this is:
- Dynamically pull the module name from generated `cargo.toml`

It looks like actor naming is related but separate from this bug -- this is a problem with the Rust compilation code not controlling the build process explicitly enough and trying to guess, essentially. It can just read from what is generated.

I'll have an updated PR for this before long.

There's still a mismatch in the what actor names are expected (theoretically... we could have an actor that just has a single emoji for a name, right?), but the immediate problem is definitely that we try to guess the name for Rust.

<details>
<summary><strong>Worth saying, but less important than the above</strong></summary>

**Dispatch based on language**

> We have to do that on some level in the control flow because Rust and Go don't have the same compiler. That's why we split up them up there, and then we just hand the path of the compiled WASM actor to the rest of the function.

Rust and Golang also don't have the same naming rules :wink: . We haven't found any bugs in the TinyGo implementation because we haven't happened upon the names that break the compiler options we're passing *yet*. 

This interestingly brings up another point -- it's weird that `wash` *invokes the compiler* at all. Pushing language-specific work/building/fiddling out of `wash` by having the template provide a consistent build point seems like a better approach. AFAIK This is how most composable multi-ecosystem build systems work (ex. buildpacks, bazel) -- rather than reaching "in" to *do* the compilation, it's completely contained in the extension point (in our case templates), and the extension provides a standardized thing.

Sounds like more fodder for an RFC :smiley: . 

Getting back to the task at hand though -- IMO what we're giving up on here is certainty while telling someone what the compiled WASM will be named. We have control over this for TinyGo because we reach in and do the build, but the place where we don't (Rust), we don't have control, and thus the bug.

Rehashing a bit: If we make a name that is known good for the compiler and give that to the template, we can have some certainty, but if we don't then we must either run their build (and specify the name) or read their build dir for the answer (searching cargo.toml).

**Transforming names at wash level or template level**

> I don't think it is difficult for a template author to write {{to_camel_case ...}} or similar when they want to transform the name. I think that also allows for greater flexibility when creating templates, and makes it clear to the users of the template what is happening. 

True, it's not too difficult, but IMO the problem is them knowing they'd have to. They don't *want* to transform the name, they have to transform it to match the rules of their language, after having been given a name that could be anything (it could be a string of emojis!). 

So the work in the template is not *just* `to_camel_case`, it's any other diff between what actor names allow and what their platform allows -- in the simpler cases this is `to_camel_case`, and in more complex cases it's `convert_emoji` or `convert_utf8`.

To be fair the work is only done once, so maybe it's negligible.

> Also I don't think cargo is a good comp because it doesn't deal with multiple language ecosystems. Something like [Bazel](https://bazel.build/) or another cross-language build toolchain may be useful to compare here.

As far as I can see [Bazel errs on the side of leaving the option in](https://bazelbuild.github.io/rules_rust/cargo.html). While Bazel does have [a strict naming scheme for packages](https://bazel.build/concepts/labels#package-names) (this is great fodder for the RFC), at the language level the default *isn't* the Bazel package name, it's what it expects rust to handle. In fact Bazel's package naming rules definitely allow for invalid [crate names](https://doc.rust-lang.org/cargo/reference/manifest.html#the-name-field) -- rather than pick the least common overlap, it seems that Bazel just side-steps the problem entirely by requiring it to be set explicitly.

I haven't used Bazel for anything substantial so I'd really appreciate corrections in my understanding :bow:  (I really want to use it in the future).

**Templates**

Hmnn I'm conflicted because I agree with you on the minimal setup point, and things building out of the box, but that power (and responsibility) should lie with template *writers*. 

Giving template writers power/flexibility to build something that gets it right without cruft/tweaking seems like a win. If we want to enforce quality on template builders, then we can easily *test* their templates before marking them with some quality indicator.

As a template writer, I'd be less concerned with `wash` wanting a minimal repo and more concerned with creating the repo that has the code I want to provide baked in given what I know the user wants/has specified.

If `wash` chooses not to carry user preferences through to me, then my hands are tied but that doesn't really spark joy. I either:

1. Write comments/README/etc for every customization point
2. Write custom templates that are slightly different for each case (`example-x` vs `example-y`)

I do understand that your stance is that (1) is a good way forward there, and there's no one clamoring for this yet, so that's the last I'll say on that to avoid taking up any more of your time!

</details>

**Comment by jordan-rash** [comment]
My 2 cents 

> I'm seeing it as setting up the minimal things needed to build a wasmCloud actor. In my eyes, 100% of the time the user should be able to create a new actor from the template and then immediately run wash build and see it working.

💯 this.  IMO, templates (that are provided by wash out of the box) are a bare minimum, functional, actor or provider that use standard practices for the selected language. 

> that power (and responsibility) should lie with template writers.

Also this.  If a user wants something more then the minimum, we have to give them the power to do that.  I think of `helm` charts at this point.  They give me the power to write my own handlebar shortcodes and if i want something special, I take it from there.  

>I don't think it is difficult for a template author to write {{to_camel_case ...}} or similar when they want to transform the name. I think that also allows for greater flexibility when creating templates, and makes it clear to the users of the template what is happening

Agree. I do not think that we should make any hard decisions on how things are or are not named.  We start out with opinions based on language specific paradigms and let the enduser take it from there.  I have a company with a certain naming convention for artifacts, if I have to go rename everything after is comes out the back of a code/template generator, I am less inclined to use that generator and I just hand role things.  

## TLDR:
Implement language best practices, give the end user the power to change what they want, provide bare minimum (but compilable) examples as `wash` defaults.

**Comment by ghost** [comment]
> But it looks like what I should actually to solve this is:
> Dynamically pull the module name from generated cargo.toml

Yes - to fix this particular bug we just need to read the Cargo.toml crate name and then use that when grabbing the compiled WASM output. #333 might be useful as a reference for that.

For the naming convention stuff I opened an RFC over [here](https://github.com/wasmCloud/wasmCloud/issues/271), I think I've derailed this issue enough over this topic lol.

I think we should also move the templates stuff out of here. @vados-cosmonic if you want to create a new issue/RFC with your thoughts on how we should change templates / the build process I'd be happy to read it!

**Comment by andar1an** [comment]
Just as a user note :

I have converted all naming in my Rust-based projects to_snake_case (this also aligns better with rust web), and some include number5_in_name which was desired by me and I had to go through and manually change names because of the out-of-the-box number handling in a name (e.g. oauth_2 to oauth2).  

One of the friction points for me entering the ecosystem was seeing a mix of dashes and underscores and predefined number handling notions within the project's various points of name representation, and this resulted in significant wasted time in determining why things wouldn't work at points. 

I am not sure how a lowest common denominator for naming could be investigated for multiple languages, but I do believe that mixed naming formats are highly tedious for any language, and a solution decision may consider this such that for a specific language, naming is consistent wherever it is represented.  

It may also be nice to know as a user where the points of representation should be verified per language as there are many files and blocks of config that display a name. This part may relate to versions as well (e.g. how wasmcloud.toml version is used when cargo.toml version is not - how does this sync easily for a user in various languages?)

**Comment by vados-cosmonic** [comment]
@mattwilkinsonn 

Appreciate the pointer and the lively discussion, giving that RFC a read!

@stephenandary 

Thanks for the input! Let me try and filter the discussion so far to what you've run into:

> One of the friction points for me entering the ecosystem was seeing a mix of dashes and underscores and predefined number handling notions within the project's various points of name representation, and this resulted in significant wasted time in determining why things wouldn't work at points.

I'd love to hear more about this -- if you could point me to somewhere we use dashes *and* underscores in a name I can definitely try and get rid of that (that really should be an bug/issue).

> I am not sure how a lowest common denominator for naming could be investigated for multiple languages, but I do believe that mixed naming formats are highly tedious for any language, and a solution decision may consider this such that for a specific language, naming is consistent wherever it is represented.

Agreed! However consensus so far seems to be that language *specific* fixes aren't the way going forward, [see the RFC](https://github.com/wasmCloud/wasmCloud/issues/271).

It *seems* like we could get away with that lowest common denominator (and it would happen to fix the rust case).

> It may also be nice to know as a user where the points of representation should be verified per language as there are many files and blocks of config that display a name. This part may relate to versions as well (e.g. how wasmcloud.toml version is used when cargo.toml version is not - how does this sync easily for a user in various languages?)

Maybe it makes sense to write some documentation on the plugin/template flow and have it be more concretely spelled out somewhere or more prominently. 

Adding documentation (FAQ/Cookbook at least) could make the contract between `wash` and the templates that are being used really obvious (and answer questions of versions). Also makes it clear what's supposed to happen versus what does.

**Comment by andar1an** [comment]
@mattwilkinsonn 
I think you misunderstood what I was saying. It is not about a mix in a single case of -'s and _'s, it is about some names using dashes and some using underscores. E.g. A project name could be `some-project`, then the binary name could be `some_project`. When you factor in numbers that can become `some2-project` and `some2_project` and `some_2_project`.  (Makefiles, cargo.toml, wasm binaries, wasmcloud.toml, and test files are some of the places a project name is used, and there are often multiple representations in a single file for binary, directory, or project name). 

I am not saying that there should be language-specific solutions, I am saying that whatever solution is decided upon should try to be as consistent as possible with naming wherever there is name representation (I don't really care in the discussion of _'s vs -'s, but from my experience in wasmcloud _'s would be the better choice). I also personally think it is better to be divergent with language naming if it means consistency within a project. 

When I am referring to a lowest common denominator, I mean settling on a naming structure that satisfies the most languages out of the gate, and then being able to convert names for other languages. snake_case seems to satisfy a lot - https://simple.wikipedia.org/wiki/Snake_case - especially of foundational languages in wasmcloud. Hopefully, by categorizing languages, the mapping of snake_case to CamelCase et al. becomes easier to post-process. and it would not be language-specific but naming convention specific. Kind of like postprocessing g-code for a specific brand of machine - have a matrix of languages that require a specific post process for names.

**Comment by vados-cosmonic** [comment]
@stephenandary I see what you meant now, it looks like I misunderstood! 

The good news is your points are [basically all covered in the RFC](https://github.com/wasmCloud/wasmCloud/issues/271), I think it's got just about everything you're hoping for, so your comments would be greatly appreciated there as well.

**Comment by andar1an** [comment]
Oops @vados-cosmonic , I @'d the wrong person in previous lol, it is late. I will look at the RFC tomorrow, but I am not sure if the link to rfc is correct. When I click it is just bring me back to this issue I created. Would it be in wasmCloud repo? Or is the RFC this issue?

**Comment by vados-cosmonic** [comment]
Hey @stephenandary Ah gotcha -- I saw that the link was back to this PR (meant to point to the comment from Matt earlier!), I've updated the links now, but just in case:

https://github.com/wasmCloud/wasmCloud/issues/271

**Comment by connorsmith256** [comment]
Resolved by #390
